### PR TITLE
Rename `Val` `evaluate` to `resolve` and implement viewport variant support

### LINKED ISSUE #9535: Update the `Val::evaluate` function to support viewport coordinates
## What problem does this solve or what need does it fill?

The `evaluate` method of `Val` needs to be updated to support `Val`'s viewport variants.

## What solution would you like?

`Val` `evaluate` and all of `Val`'s `_with_size` functions that use `evaluate` will need a viewport size parameter.

[Viewport coordinates are explained here](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units). 

The tests also need to be updated and expanded to cover the viewport variants.

### PR DESCRIPTION
**Pull request by ickshonpe**
# Objective

Rename `Val`'s `evaluate` method to `resolve`.

Implement `resolve` support for `Val`'s viewport variants.

fixes #9535

---

## Changelog

`bevy_ui::ui_node::Val`:
* Renamed the following methods and added a `viewport_size` parameter:
   - `evaluate` to `resolve`
   - `try_add_with_size` to `try_add_with_context`
   - `try_add_assign_with_size` to `try_add_assign_with_context`
   - `try_sub_with_size` to `try_sub_with_context`
   - `try_sub_assign_with_size` to `try_sub_assign_with_context`
* Implemented `resolve` support for `Val`'s viewport coordinate types

## Migration Guide
* Renamed the following `Val` methods and added a `viewport_size` parameter:
   - `evaluate` to `resolve`
   - `try_add_with_size` to `try_add_with_context`
   - `try_add_assign_with_size` to `try_add_assign_with_context`
   - `try_sub_with_size` to `try_sub_with_context`
   - `try_sub_assign_with_size` to `try_sub_assign_with_context`

### CODE REVIEW

**Comment by ickshonpe** [comment]
Maybe both `evaluate` and `eval` are both wrong, `resolve` seems more idiomatic.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
`resolve` is a much better name. Code changes look good though!

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ui/src/ui_node.rs [hunk 1]
```diff
@@ -238,15 +238,19 @@ pub enum ValArithmeticError {
 
 impl Val {
     /// A convenience function for simple evaluation of [`Val::Percent`] variant into a concrete [`Val::Px`] value.
```
Comment (nicoburns) [inline crates/bevy_ui/src/ui_node.rs:240]: This line should probably be updated to reflect that it is not just the `Val::Percent` variant that is resolved.

File: crates/bevy_ui/src/ui_node.rs [hunk 2]
```diff
@@ -238,15 +238,19 @@ pub enum ValArithmeticError {
...
-    /// Returns a [`ValArithmeticError::NonEvaluateable`] if the [`Val`] is impossible to evaluate into [`Val::Px`].
-    /// Otherwise it returns an [`f32`] containing the evaluated value in pixels.
+    /// Returns a [`ValArithmeticError::NonEvaluateable`] if the [`Val`] is impossible to resolve into [`Val::Px`].
+    /// Otherwise it returns an [`f32`] containing the resolved value in pixels.
     ///
-    /// **Note:** If a [`Val::Px`] is evaluated, it's inner value returned unchanged.
-    pub fn evaluate(&self, size: f32) -> Result<f32, ValArithmeticError> {
+    /// **Note:** If a [`Val::Px`] is resolved, it's inner value is returned unchanged.
+    pub fn resolve(self, size: f32, viewport_size: Vec2) -> Result<f32, ValArithmeticError> {
```
Comment (nicoburns) [inline crates/bevy_ui/src/ui_node.rs:245]: Perhaps the `size` parameter should be renamed to `percentage_size` or `parent_size`? As `viewport_size` is also a `size`...

Comment (nicoburns) [reply crates/bevy_ui/src/ui_node.rs:245]: Could even call this `resolve_to_px` if we wanted to be really explicit (although I think `resolve` is fine).

**Review by nicoburns — APPROVED** [review]
Looks good to me. I've left a few suggestions around docs/naming.

File: crates/bevy_ui/src/ui_node.rs (hunk 2 above)
Comment (ickshonpe) [reply crates/bevy_ui/src/ui_node.rs:245]: Completely forgot about updating the docs for this one.

Comment (ickshonpe) [reply crates/bevy_ui/src/ui_node.rs:245]: `parent_size` is ok, but it doesn't feel super satisfying. Maybe `parent_length` might be better? We've mostly only used size to refer to 2d properties in the past.

Comment (ickshonpe) [reply crates/bevy_ui/src/ui_node.rs]: `parent_size` is good enough for now, can always be changed later.

**Pull request merged**